### PR TITLE
GIT-3100: Reconsidered the password complexity check design.

### PR DESCRIPTION
**Pull request by KH-Amir-TN**
## Description

As of the latest release of Greenlight (GL) `v2.11.x` the password complexity has increased.
The solution design used a flag (`secure_password`) to mark accounts with "secure/insecure" passwords so that the application can decide the proper action to take with each account.
Any user local account with insecure password (`secure_password == false`) won't login and gets redirected directly to `reset_password` page to update it matching the new requirements.
This raised some issues as described in #3078,#3079,#3081, #3091,#3100.
This addresses #3100 and #3091 while PR #3082 targets the others.

## Solution:
This proposition is straightforward, any local account will have its password validated against the complexity requirements upon login without the use of the `secure_password` flag which simplifies things a lot and avoid potential errors and overhead:
1. If the password is secure then the user login.
2. If the password is insecure (doesn't match the `PASSWORD_PATTERN`) the user won't login and they will get redirected to reset password page.
This will ensure that old accounts with already "secure" passwords doesn't get prompted to change theirs.

The password minimal complexity required remains the same:
1. A password has to have at least **One digit**.
2. A password has to have at least **One non-word characters (anything but `a-z, A-Z, 0-9`)**.
3. A password has to have at least **One Upper case letter**.
4. A password has to have at least **One lower case letter**.
5. A password has to be at least **8 characters in length**.

## Testing Steps

1. Ran `bigbluebutton/greenlight:v2.10.0.3` on empty database.
2. Created three accounts two of which with insecure and one with a secure password.
3. Opened three browser sessions for each account.
4. Ran this PR docker image on the existing database.
5. Attempted to change the password on one account with an insecure one:
   5.1 Confirmed the complexity validation for the new password.
   5.2 Changed the password to a secure one.
   5.3 Signed out and resigned in on same account and confirming that the user doesn't has to reset their password.
6. Signed out all of the other sessions and resigned in to confirm that:
   6.1 The account with insecure password can't login and gets redirected to the reset password page:
       6.1.1 All validation for password complexity works.
       6.1.2 After resetting the password and signing in the user logs in without having to reset it again.
       6.1.3 Signing out and resigning in without having to reset the password.
   6.2 The user with secure password logs in without having to make a reset.

Also, Tested for `user` and `admin` create rake tasks.

### CODE REVIEW

File: db/migrate/20211128175028_add_secure_password_to_users.rb [hunk 1]
```diff
@@ -1,7 +0,0 @@
-# frozen_string_literal: true
-
-class AddSecurePasswordToUsers < ActiveRecord::Migration[5.2]
-  def change
-    add_column :users, :secure_password, :boolean, null: false, default: false
-  end
-end
```
Comment (farhatahmad) [inline db/migrate/20211128175028_add_secure_password_to_users.rb:7]: Deleting a migration will cause other issues. You should instead create a new migration that removes the column instead

File: app/models/user.rb [hunk 2]
```diff
@@ -114,6 +115,10 @@ def merge_list_search(string)
       search_param = "%#{sanitize_sql_like(string)}%"
       where(search_query, search: search_param)
     end
+
+    def secure_password?(pwd)
+      PASSWORD_PATTERN === pwd
```
Comment (farhatahmad) [inline app/models/user.rb:120]: Hmmm I've never seen `===` used to check regex. I normally would use
`pwd.match(PASSWORD_PATTERN)`

**Comment by farhatahmad** [comment]
Other than the 2 issues above, this looks good. I'll merge this and look at a new release once the issues are fixed

File: db/migrate/20211128175028_add_secure_password_to_users.rb (hunk 1 above)
Comment (KH-Amir-TN) [reply db/migrate/20211128175028_add_secure_password_to_users.rb:7]: You got reason it's wrong.
I'll change it.

File: app/models/user.rb (hunk 2 above)
Comment (KH-Amir-TN) [reply app/models/user.rb:120]: It's a case equality that `case` uses under the hood.
https://ruby-doc.org/core-3.0.2/Regexp.html#method-i-3D-3D-3D
Do you want me to use `#string.match`?

Comment (KH-Amir-TN) [reply app/models/user.rb]: DONE

File: db/migrate/20211128175028_add_secure_password_to_users.rb (hunk 1 above)
Comment (KH-Amir-TN) [reply db/migrate/20211128175028_add_secure_password_to_users.rb]: DONE

**Review by farhatahmad — APPROVED** [review]

**Pull request merged**